Sprint 7

bakgeldia opened the pull request:
Задача 7-спринта.
- Написаны unit-тесты
- Написаны UI-тесты
- Проведён рефакторинг проекта — отрефакторен MovieQuizViewController на архитектуру MVP
- Написан и проходит тест для convert(model: QuizQuestion) -> QuizStepViewModel.

bakgeldia committed:
Добавлены:
- Unit-тесты
- UI-тесты

bakgeldia committed:
Рефакторинг кода (из MVC в MVP):
- Был создан MovieQuizPresenter
- Метод convert(model: QuizQuestion) был перенесен в MovieQuizPresenter
- Переменные questionsAmount и currentQuestionIndex также были перенесены в MovieQuizPresenter
- Добавлено свойство presenter в MovieQuizViewControler
- Применен DRYдля устранения дублирования данных.

bakgeldia committed:
Рефакторинг:
- Логика yesButtonClicked(_ sender: UIButton) перенесена в presenter
- Логика noButtonClicked(_ sender: UIButton) перенесена в presenter

bakgeldia committed:
Рефакторинг:
- DRY: устранено дублирование, возникшее после перенесении логик yesButtonClicked(_ sender: UIButton) и noButtonClicked(_ sender: UIButton) в presenter;
- DRY: устранено дублирование в использовании currentQuestion после логик yesButtonClicked(_ sender: UIButton) и noButtonClicked(_ sender: UIButton) в presenter;
- Метод showNextQuestionOrResults перенесен в presenter;
- Применен DRY: устранить дублирование correctAnswers;
- Применен DRY: устранить дублирование questionFactory;
- Логика формирования сообщения с результатами квиза перенесен в метод presenter'а makeResultsMessage(), а также применен DRY для устранения дублирования свойства statisticService;
- Асинхронный код (DispatchQueue.main.asyncAfter) перенесен в presenter

bakgeldia committed:
Настройки уровня доступа private для полей и методов в Presenter, которые не используются напрямую из ViewController

bakgeldia committed:
Тестирование Presenter'а

So1ovyov commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -38,3 +40,5 @@ extension AlertPresenterImplementation: AlertPresenter {
 }
 
+
+//alert.view.accessibilityIdentifier = "Game results"
```
Закомментированный код забыл удалить

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,144 @@
+
+final class MovieQuizPresenter: QuestionFactoryDelegate {
+    
+    private let statisticService: StatisticService!
```
Давай обезопасим проект и поменяем с force на опционал `private let statisticService: StatisticService?`

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,144 @@
+        return newModel
+    }
+    
+    func yesButtonClicked() {
```
А почему бы по аналогии не создать универсальный метод для ответа Да/Нет:
```
func buttonClicked(isYes: Bool) {
   didAnswer(isYes: isYes)
}
```

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -11,56 +11,24 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
-    
     private var alertPresenter: AlertPresenter?
-    private var statisticService: StatisticService?
+    private var presenter: MovieQuizPresenter!
```
Аналогично и здесь предлагаю поменять на опционал

So1ovyov commented on MovieQuiz/Services/MoviesLoader.swift:
```diff
@@ -39,7 +41,7 @@ struct MoviesLoader: MoviesLoading {
-    
 }
+
+//https://tv-api.com/en/API/Top250Movies/k_zcuw1ytf
```
Не забудь и здесь почистить закомментированный код

So1ovyov commented on MovieQuizPresenterTests/MovieQuizPresenterTests.swift:
```diff
@@ -0,0 +1,51 @@
+
+// Mock ViewController
+final class MovieQuizViewControllerMock: MovieQuizViewControllerProtocol {
+    func show(quiz step: QuizStepViewModel) {
```
Предлагаю не раздувать файл и пустые методы написать друг за другом:
```
    func show(quiz step: QuizStepViewModel) {}
    func showFinalResult(quiz result: QuizResultsViewModel) {}
    func highlightImageBorder(isCorrectAnswer: Bool) {}
```

So1ovyov commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,36 @@
+import XCTest // не забывайте импортировать фреймворк для тестирования
+@testable import MovieQuiz // импортируем наше приложение для тестирования
+
+class ArrayTests: XCTestCase {
```
Проверь во всех классах наличие модификатора final, он не нужен только если ты планируешь наследоваться от класса

So1ovyov commented on MovieQuizTests/ArrayTests.swift:
```diff
@@ -0,0 +1,36 @@
+//
+
+import Foundation
+import XCTest // не забывайте импортировать фреймворк для тестирования
```
Комментарии с советами можно удалить

So1ovyov commented on MovieQuizTests/MoviesLoaderTests.swift:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+struct StubNetworkClient: NetworkRouting {
```
Давай придерживаться принципа 1 файл = 1 структура/класс/модель/enum/расширение. Не стоит раздувать файл и располагать в нем все сразу.

bakgeldia committed:
Исправления:
- Удалены лишние комментарии
- Безопасный unwrapping
- Данные не сохранялись в устройстве, поэтому и не показывались в конце квиза. Исправлено, добавлением кода  statisticService.store(correct: correctAnswers, total: questionsAmount) в MovieQuizPresenter
- Добавлен модификатор final для классов, от которых не наследуются.
